File: android/app/src/main/java/com/dayforge/domain/model/ConfigContract.kt
@file:kotlinx.serialization.UseSerializers(ContractStringSerializer::class)

package com.dayforge.domain.model

import java.time.LocalDate
import java.time.ZoneId
import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonClassDiscriminator

@Serializable
@OptIn(ExperimentalSerializationApi::class)
@JsonClassDiscriminator("type")
sealed interface ConfigSchedule {
    @Serializable @SerialName("daily")
    data class Daily(@SerialName("start_date") val startDate: String?) : ConfigSchedule {
        init { require(isConfigDay(startDate)) }
    }

    @Serializable @SerialName("weekly")
    data class Weekly(
        @SerialName("start_date") val startDate: String?,
        val weekdays: List<@Serializable(with = ContractIntegerSerializer::class) Int>
    ) : ConfigSchedule {
        init {
            require(isConfigDay(startDate))
            require(weekdays.isNotEmpty() && weekdays.size <= 7)
            require(weekdays.all { it in 1..7 } && weekdays.distinct().size == weekdays.size)
        }
    }

    @Serializable @SerialName("monthly")
    data class Monthly(
        @SerialName("start_date") val startDate: String?,
        @Serializable(with = ContractIntegerSerializer::class)
        @SerialName("day_of_month") val dayOfMonth: Int
    ) : ConfigSchedule {
        init { require(isConfigDay(startDate) && dayOfMonth in 1..31) }
    }

    @Serializable @SerialName("interval")
    data class Interval(
        @SerialName("start_date") val startDate: String,
        @Serializable(with = ContractIntegerSerializer::class)
        @SerialName("every_days") val everyDays: Int
    ) : ConfigSchedule {
        init { require(isConfigDay(startDate) && everyDays in 1..3650) }
    }

    @Serializable @SerialName("once")
    data class Once(@SerialName("due_date") val dueDate: String?) : ConfigSchedule {
        init { require(isConfigDay(dueDate)) }
    }
}

@Serializable
data class ConfigAppearance(
    val icon: IconReference,
    @SerialName("accent_color") val accentColor: String,
    @SerialName("icon_tint") val iconTint: String
) {
    init {
        require(isAccentColor(accentColor))
        require(iconTint == "theme" || iconTint == "object")
    }
}

@Serializable
data class ConfigActivity(
    @SerialName("tracking_mode") val trackingMode: String,
    @SerialName("completion_policy") val completionPolicy: String,
    @Serializable(with = ContractBooleanSerializer::class)
    @SerialName("is_countdown") val isCountdown: Boolean,
    @Serializable(with = ContractIntegerSerializer::class)
    @SerialName("target_value") val targetValue: Int,
    @Serializable(with = ContractIntegerSerializer::class)
    @SerialName("target_cycles") val targetCycles: Int?,
    @SerialName("fail_mode") val failMode: String,
    @Serializable(with = ContractIntegerSerializer::class)
    @SerialName("preferred_minute") val preferredMinute: Int?,
    val timezone: String,
    val schedule: ConfigSchedule
) {
    init {
        require(trackingMode in setOf("check", "count", "duration"))
        require(completionPolicy in setOf("recurring", "one_and_done"))
        require(failMode == "strict" || failMode == "loose")
        require(targetValue >= 0 && (targetCycles == null || targetCycles > 0))
        require(preferredMinute == null || preferredMinute in 0..1439)
        require(timezone.length in 1..64 && timezone in ZoneId.getAvailableZoneIds())
        val once = completionPolicy == "one_and_done"
        require(once == (schedule is ConfigSchedule.Once))
        require(trackingMode != "check" || !isCountdown)
        require(trackingMode == "check" || targetValue > 0)
        require(trackingMode != "duration" || (targetValue % 60 == 0 && targetValue <= 2_147_483_640))
        require(!once || (trackingMode == "check" && targetValue == 1 && targetCycles == null &&
            failMode == "loose" && preferredMinute == null))
    }
}

@Serializable
data class ConfigGoal(
    @SerialName("start_date") val startDate: String?,
    @SerialName("due_date") val dueDate: String?,
    @Serializable(with = ContractIntegerSerializer::class)
    @SerialName("target_cycles") val targetCycles: Int?,
    @SerialName("fail_mode") val failMode: String
) {
    init {
        require(isConfigDay(startDate) && isConfigDay(dueDate))
        require(startDate == null || dueDate == null || startDate <= dueDate)
        require(targetCycles == null || targetCycles > 0)
        require(failMode == "strict" || failMode == "loose")
    }
}

@Serializable
data class ConfigNode(
    val key: String,
    val kind: String,
    val name: String,
    val description: String,
    @Serializable(with = ContractBooleanSerializer::class)
    @SerialName("is_active") val isActive: Boolean,
    @SerialName("parent_key") val parentKey: String?,
    val appearance: ConfigAppearance,
    val goal: ConfigGoal?,
    val activity: ConfigActivity?
) {
    init {
        require(isLocalKey(key) && (parentKey == null || isLocalKey(parentKey)))
        require(isContractName(name, 100) && description.codePointCount(0, description.length) <= 1000)
        require(kind == "goal" || kind == "activity")
        require(if (kind == "goal") parentKey == null && goal != null && activity == null
            else activity != null && goal == null)
    }
}

@Serializable
data class ConfigMetric(
    val key: String,
    val name: String,
    val description: String,
    val unit: String,
    @Serializable(with = ContractBooleanSerializer::class)
    @SerialName("is_active") val isActive: Boolean,
    @Serializable(with = ContractIntegerSerializer::class)
    @SerialName("decimal_places") val decimalPlaces: Int,
    @SerialName("aggregation_type") val aggregationType: String,
    @SerialName("target_direction") val targetDirection: String?,
    @Serializable(with = ContractNumberSerializer::class)
    @SerialName("target_value") val targetValue: Double?,
    @Serializable(with = ContractNumberSerializer::class)
    @SerialName("target_value_upper") val targetValueUpper: Double?,
    val appearance: ConfigAppearance
) {
    init {
        require(isLocalKey(key) && isContractName(name, 100) && isContractName(unit, 50))
        require(description.codePointCount(0, description.length) <= 1000 && decimalPlaces in 0..6)
        require(aggregationType in setOf("average", "sum", "by_time"))
        require(targetDirection == null || targetDirection in setOf("increase", "decrease", "range"))
        require(targetValue == null || targetValue.isFinite())
        require(targetValueUpper == null || targetValueUpper.isFinite())
        require(targetDirection != "range" || (targetValue != null && targetValueUpper != null && targetValueUpper >= targetValue))
    }
}

@Serializable
data class ConfigLink(
    val key: String,
    @SerialName("activity_key") val activityKey: String,
    @SerialName("metric_key") val metricKey: String,
    @Serializable(with = ContractNumberSerializer::class)
    val coefficient: Double,
    @Serializable(with = ContractBooleanSerializer::class)
    @SerialName("show_in_detail") val showInDetail: Boolean,
    @Serializable(with = ContractBooleanSerializer::class)
    @SerialName("prompt_on_complete") val promptOnComplete: Boolean,
    @Serializable(with = ContractBooleanSerializer::class)
    @SerialName("is_active") val isActive: Boolean
) {
    init { require(isLocalKey(key) && isLocalKey(activityKey) && isLocalKey(metricKey) && coefficient.isFinite()) }
}

/** File-local identities are references only, never authorization to upsert an account's existing data. */
@Serializable
data class ConfigBundle(
    val format: String,
    @Serializable(with = ContractIntegerSerializer::class)
    @SerialName("format_version") val formatVersion: Int,
    val nodes: List<ConfigNode>,
    val metrics: List<ConfigMetric>,
    val links: List<ConfigLink>,
    @SerialName("icon_pack") val iconPack: IconPack?,
    @SerialName("unresolved_roles") val unresolvedRoles: List<String>,
    val themes: List<ThemeDefinition>
) {
    init {
        require(format == "dayforge.config" && formatVersion == 2)
        require(nodes.size <= 1000 && metrics.size <= 1000 && links.size <= 5000)
        require(unresolvedRoles.size <= 256 && themes.size <= 16)
        val keys = nodes.map { it.key } + metrics.map { it.key } + links.map { it.key }
        require(keys.distinct().size == keys.size)
        require(nodes.map { it.name }.distinct().size == nodes.size)
        require(metrics.map { it.name }.distinct().size == metrics.size)
        val byKey = nodes.associateBy { it.key }
        val metricKeys = metrics.map { it.key }.toSet()
        require(nodes.all { it.parentKey == null || byKey[it.parentKey]?.kind == "goal" })
        require(links.map { it.activityKey to it.metricKey }.distinct().size == links.size)
        require(links.all { byKey[it.activityKey]?.kind == "activity" && it.metricKey in metricKeys })
        val assets = iconPack?.assets?.associateBy { it.assetId }.orEmpty()
        val roles = iconPack?.roles.orEmpty()
        val unresolved = unresolvedRoles.toSet()
        require(unresolved.size == unresolvedRoles.size && unresolved.intersect(roles.keys).isEmpty())
        unresolved.forEach { IconReference.Role(it) }
        val usedRoles = mutableSetOf<String>()
        fun validateIcon(icon: IconReference, once: Boolean) {
            when (icon) {
                is IconReference.Role -> {
                    usedRoles += icon.role
                    require(icon.role in roles || icon.role in unresolved)
                    require(iconAllowed(icon, once))
                }
                is IconReference.Asset -> require(iconAllowed(icon, once, assets[icon.assetId]))
            }
        }
        nodes.forEach { validateIcon(it.appearance.icon, it.activity?.completionPolicy == "one_and_done") }
        metrics.forEach { validateIcon(it.appearance.icon, false) }
        require(unresolved.all { it in usedRoles })
        require(themes.map { it.themeId to it.revision }.distinct().size == themes.size)
    }
}

private val localKey = Regex("[a-z][a-z0-9_-]{0,63}")
private val configDay = Regex("[0-9]{4}-[0-9]{2}-[0-9]{2}")
private fun isLocalKey(value: String): Boolean = localKey.matches(value)
private fun isConfigDay(value: String?): Boolean = value == null ||
    (configDay.matches(value) && runCatching { LocalDate.parse(value).year in 1..9999 }.getOrDefault(false))

File: backend/src/v2/config_contract.py
"""Portable config-v2 metadata. Import remaps every package identity; no history."""

from datetime import date
from typing import Annotated, Literal

from pydantic import Field, field_validator, model_validator

from src.v2.appearance import (
    AssetIcon,
    IconPack,
    IconReference,
    RoleIcon,
    RoleKey,
    icon_allowed,
)
from src.v2.contract_types import (
    AccentColor,
    ContractModel,
    exact_integer,
    visible_name,
)
from src.v2.theme_contract import ThemeDefinition
from src.v2.time_utils import require_iana_timezone


LocalId = Annotated[str, Field(pattern=r"^[a-z][a-z0-9_-]{0,63}$")]
Day = Annotated[str, Field(pattern=r"^[0-9]{4}-[0-9]{2}-[0-9]{2}$")]


class DatedSchedule(ContractModel):
    start_date: Day | None

    @field_validator("start_date")
    @classmethod
    def valid_day(cls, value):
        if value is not None:
            date.fromisoformat(value)
        return value


class DailySchedule(DatedSchedule):
    type: Literal["daily"]


class WeeklySchedule(DatedSchedule):
    type: Literal["weekly"]
    weekdays: list[Annotated[int, Field(ge=1, le=7)]] = Field(
        min_length=1, max_length=7
    )

    @model_validator(mode="after")
    def unique_days(self):
        if len(set(self.weekdays)) != len(self.weekdays):
            raise ValueError("duplicate weekdays")
        return self


class MonthlySchedule(DatedSchedule):
    type: Literal["monthly"]
    day_of_month: int = Field(ge=1, le=31)


class IntervalSchedule(DatedSchedule):
    type: Literal["interval"]
    start_date: Day
    every_days: int = Field(ge=1, le=3650)


class OnceSchedule(ContractModel):
    type: Literal["once"]
    due_date: Day | None

    @field_validator("due_date")
    @classmethod
    def valid_day(cls, value):
        if value is not None:
            date.fromisoformat(value)
        return value


ConfigSchedule = Annotated[
    DailySchedule | WeeklySchedule | MonthlySchedule | IntervalSchedule | OnceSchedule,
    Field(discriminator="type"),
]


class ConfigAppearance(ContractModel):
    icon: IconReference
    accent_color: AccentColor
    icon_tint: Literal["theme", "object"]


class ConfigActivity(ContractModel):
    tracking_mode: Literal["check", "count", "duration"]
    completion_policy: Literal["recurring", "one_and_done"]
    is_countdown: bool
    target_value: int = Field(ge=0, le=2_147_483_647)
    target_cycles: int | None = Field(ge=1, le=2_147_483_647)
    fail_mode: Literal["strict", "loose"]
    preferred_minute: int | None = Field(ge=0, le=1439)
    timezone: str = Field(min_length=1, max_length=64)
    schedule: ConfigSchedule

    @field_validator("timezone")
    @classmethod
    def valid_timezone(cls, value):
        return require_iana_timezone(value)

    @model_validator(mode="after")
    def valid_mode(self):
        once = self.completion_policy == "one_and_done"
        if once != (self.schedule.type == "once"):
            raise ValueError("completion policy and schedule disagree")
        if self.tracking_mode == "check" and self.is_countdown:
            raise ValueError("check activities cannot count down")
        if self.tracking_mode != "check" and self.target_value <= 0:
            raise ValueError("count and duration targets must be positive")
        if self.tracking_mode == "duration" and (
            self.target_value % 60 != 0 or self.target_value > 2_147_483_640
        ):
            raise ValueError("duration target must be whole minutes in seconds")
        if once and (
            self.tracking_mode != "check"
            or self.target_value != 1
            or self.target_cycles is not None
            or self.fail_mode != "loose"
            or self.preferred_minute is not None
        ):
            raise ValueError("one-time items only support a single unplanned check")
        return self


class ConfigGoal(ContractModel):
    start_date: Day | None
    due_date: Day | None
    target_cycles: int | None = Field(ge=1, le=2_147_483_647)
    fail_mode: Literal["strict", "loose"]

    @model_validator(mode="after")
    def valid_dates(self):
        start = date.fromisoformat(self.start_date) if self.start_date else None
        end = date.fromisoformat(self.due_date) if self.due_date else None
        if start and end and start > end:
            raise ValueError("goal date range is reversed")
        return self


class ConfigNode(ContractModel):
    key: LocalId
    kind: Literal["goal", "activity"]
    name: str = Field(min_length=1, max_length=100)
    description: str = Field(max_length=1000)
    is_active: bool
    parent_key: LocalId | None
    appearance: ConfigAppearance
    goal: ConfigGoal | None
    activity: ConfigActivity | None

    @model_validator(mode="after")
    def valid_kind(self):
        visible_name(self.name, 100)
        if self.kind == "goal":
            if (
                self.parent_key is not None
                or self.goal is None
                or self.activity is not None
            ):
                raise ValueError(
                    "goals must be top-level and contain only goal details"
                )
        elif self.activity is None or self.goal is not None:
            raise ValueError("activities require activity details only")
        return self


class ConfigMetric(ContractModel):
    key: LocalId
    name: str = Field(min_length=1, max_length=100)
    description: str = Field(max_length=1000)
    unit: str = Field(min_length=1, max_length=50)
    is_active: bool
    decimal_places: int = Field(ge=0, le=6)
    aggregation_type: Literal["average", "sum", "by_time"]
    target_direction: Literal["increase", "decrease", "range"] | None
    target_value: float | None = Field(allow_inf_nan=False)
    target_value_upper: float | None = Field(allow_inf_nan=False)
    appearance: ConfigAppearance

    @model_validator(mode="after")
    def valid_metric(self):
        visible_name(self.name, 100)
        visible_name(self.unit, 50)
        if self.target_direction == "range" and (
            self.target_value is None
            or self.target_value_upper is None
            or self.target_value_upper < self.target_value
        ):
            raise ValueError("invalid metric target range")
        return self


class ConfigLink(ContractModel):
    key: LocalId
    activity_key: LocalId
    metric_key: LocalId
    coefficient: float = Field(allow_inf_nan=False)
    show_in_detail: bool
    prompt_on_complete: bool
    is_active: bool


class ConfigBundle(ContractModel):
    format: Literal["dayforge.config"]
    format_version: Literal[2]
    nodes: list[ConfigNode] = Field(max_length=1000)
    metrics: list[ConfigMetric] = Field(max_length=1000)
    links: list[ConfigLink] = Field(max_length=5000)
    icon_pack: IconPack | None
    unresolved_roles: list[RoleKey] = Field(max_length=256)
    themes: list[ThemeDefinition] = Field(max_length=16)

    @field_validator("format_version", mode="before")
    @classmethod
    def exact_version_type(cls, value):
        return exact_integer(value)

    @model_validator(mode="after")
    def valid_graph(self):
        keys = (
            [node.key for node in self.nodes]
            + [metric.key for metric in self.metrics]
            + [link.key for link in self.links]
        )
        if len(set(keys)) != len(keys):
            raise ValueError("all package keys must be unique")
        for items in (self.nodes, self.metrics):
            if len({item.name for item in items}) != len(items):
                raise ValueError("duplicate names")
        nodes = {node.key: node for node in self.nodes}
        metrics = {metric.key for metric in self.metrics}
        for node in self.nodes:
            if node.parent_key is not None:
                parent = nodes.get(node.parent_key)
                if parent is None or parent.kind != "goal":
                    raise ValueError("parent must reference a package goal")
        pairs = {(link.activity_key, link.metric_key) for link in self.links}
        if len(pairs) != len(self.links):
            raise ValueError("duplicate metric links")
        for link in self.links:
            linked_node = nodes.get(link.activity_key)
            if (
                linked_node is None
                or linked_node.kind != "activity"
                or link.metric_key not in metrics
            ):
                raise ValueError("invalid link endpoints")
        assets = (
            {a.asset_id: a for a in self.icon_pack.assets} if self.icon_pack else {}
        )
        roles = self.icon_pack.roles if self.icon_pack else {}
        unresolved = set(self.unresolved_roles)
        if len(unresolved) != len(self.unresolved_roles) or unresolved & roles.keys():
            raise ValueError("ambiguous unresolved role declaration")
        used_roles: set[str] = set()
        appearance_items: list[ConfigNode | ConfigMetric] = [*self.nodes, *self.metrics]
        for item in appearance_items:
            once = (
                isinstance(item, ConfigNode)
                and item.activity is not None
                and item.activity.completion_policy == "one_and_done"
            )
            icon = item.appearance.icon
            if isinstance(icon, RoleIcon):
                used_roles.add(icon.role)
                if icon.role not in roles and icon.role not in unresolved:
                    raise ValueError(
                        "role has no captured mapping or missing-role marker"
                    )
                if not icon_allowed(icon, one_time=once):
                    raise ValueError("icon purpose does not match the object")
            elif isinstance(icon, AssetIcon) and not icon_allowed(
                icon, one_time=once, asset=assets.get(icon.asset_id)
            ):
                raise ValueError("missing or wrong-purpose pinned asset")
        if not unresolved <= used_roles:
            raise ValueError("unused unresolved roles")
        theme_ids = [(theme.theme_id, theme.revision) for theme in self.themes]
        if len(set(theme_ids)) != len(theme_ids):
            raise ValueError("duplicate theme version")
        return self
